### DISCUSSION TITLE
Ability to use a different font

### DISCUSSION THREAD

**Discussion by dtcyganok** [ihhub/fheroes2#8387] [General]
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.

### Describe the problem requiring a solution

The default font is extremely illegible and unreadable. So even a few sentenced text is nearly impossible to read. Maybe this typeface was supposed to create a fantasy atmosphere. But what's the point if a person can't read it?

### Describe the possible solution

It would be nice to have an UI element to control the typeface. Or just having a config option would be nice as well.

**Comment by ihhub** [comment]
Hi @dtcyganok , what is wrong with the current font? How it can be small playing the game in full screen:
![image](https://github.com/ihhub/fheroes2/assets/19829520/76f96b47-779d-4d8b-a9f7-8f3d24502b08)

**Comment by ihhub** [comment]
I am converting this into a discussion topic.

**Comment by dtcyganok** [comment]
The font size isn't the problem. The problem is in the font appearance (the shape of the letters). Some letters look really strange. It's probably easier to show some examples:
- The uppercase `I` looks exactly as the letter `J` 
![Screenshot from 2024-02-10 01-11-25](https://github.com/ihhub/fheroes2/assets/6991483/b25f027e-c7f0-4f71-943e-e18ea3f87ac5)
- The dash symbol `-` looks strange
![Screenshot from 2024-02-10 01-18-30](https://github.com/ihhub/fheroes2/assets/6991483/e3c27635-1c11-4621-9b0e-c43a5d4bcac3)

Besides there are a lot of kerning (the distance between the letters) problems. Here are just two examples:
- The word `defeat` looks like `def eat`
![Screenshot from 2024-02-10 01-17-12](https://github.com/ihhub/fheroes2/assets/6991483/b5e7b724-d358-40e4-92a6-4505bc09bd9c)
- The word `have` looks like `ha ve`
![Screenshot from 2024-02-10 01-20-32](https://github.com/ihhub/fheroes2/assets/6991483/814bc16d-5a3e-442d-b94b-eac56dc4fcb3)

The overall font quality seems poor, so I'd prefer to use another font.

**Reply by zenseii** [reply]
What you are describing about kerning, or taking into account digraphs, is a feature that is lacking in fheroes2 but it's been discussed as something that needs to be implemented, exactly for the reasons you described.

Either way, kerning would be an issue with any other font when it's something the engine doesn't support so changing font wouldn't fix anything.

We have already made it possible to adjust symbols like the dash, and we can definitely make further adjustments to them.

The "I" can be adjusted too if you have any concrete examples of how to. We've already made adjustments to the lowercase K that you can see if you compare to the original game where it looks like it's uppercase.

**Reply by dtcyganok** [reply]
I'd say that the kerning problems are minor. And if there is some plan to implement it, that's cool.

Talking about font patching I believe that's a lot of work even for a proficient type designer. And even though we'll refine tens of symbols, many other details will left unfixed. I don't develop this project, but I'm almost certain that replacing the font completely is far more cheaper than the font patching in term of time.

**Reply by Branikolog** [reply]
Hello, @dtcyganok !
I've never heard such opinion about the original font in Heroes 2. Indeed, it's less legible, than any regular font (like "Times New Roman" or so...) But it is a fantasy game where overall atmosphere is important. I remember old fairy-tale books from my childhood, where font was about the same. It goes without saying, that the initial letter was even much less legible, being drawn in a form of some creature or subject. But it creates a certain charm. I really doubt players would be happy if we change original font into something more boring.
Maybe some minor fixes could be applied, but I personally vote against changing the font. I'm pretty satisfied with the current texts which look even weird in some cases. This creates a special ambience while running this game.

**Reply by dtcyganok** [reply]
Hi @Branikolog!

The point about the game atmosphere is perfectly correct, and I completely agree with it. 
However I definitely know there are thousands of free fonts (SIL Open Font License or something similar), and I'm pretty sure it's possible to find a really legible and atmospheric font that would suit all the needs well.

Actually I don't persist that the font should be changed. My suggestion is rather an RFC. And my main suggestion is to give a user the ability to decide which font to use.

**Reply by oleg-derevenetz** [reply]
> And my main suggestion is to give a user the ability to decide which font to use

First of all, original fonts are not TrueType fonts or something like this, their rendering process is peculiar and incompatible with modern TTF. Then, they are not Unicode fonts, because Unicode was not widely used at that time - they are 8-bit fonts for a specific national 8-bit encoding (another incompatibility with modern fonts). And also I should say that the process of enumerating available fonts is not universal (to put it mildly) and is heavily dependent on OS (or even on a specific distribution of that OS). Thus, the rework will require a lot of effort (as well as, most likely, additional dependencies), for the sake of...? Anyway, if there are those who want to implement it so that it works correctly on all supported platforms and with all supported languages, both with original and 3rd-party fonts, they are always welcome to offer a PR.

**Reply by dtcyganok** [reply]
@oleg-derevenetz thanks for your description! Now it's more clear how it works.

> for the sake of...?

For the sake of readability and pleasant UX

Seems we have two main options:

1.  Patch the font with redesigned symbols
2.  Rework all text fields in the game to support TrueType fonts

So let's get real. Only the first option achievable? Right?

**Reply by oleg-derevenetz** [reply]
> For the sake of readability and pleasant UX

So far, you are the first person who wants to add the ability to change in-game fonts to arbitrary ones. Usually games have a predefined set of fonts that match the atmosphere of the game, and do not allow you to arbitrarily change them to 3rd party fonts. Ultimately, games are not Microsoft Word.

> Only the first option achievable? Right?

I put the question differently. Any person can offer a specific working solution (at least in the form of proof-of-concept). This PoC should work on all supported platforms where fheroes2 can currently work (effectively any platform that is supported by SDL), it should support all languages which are currently supported by the game, it should support both fonts from the original game (because I foresee a lot of connoisseurs of the original game, dissatisfied with the potential forced font replacement to some 3rd party TTF) and 3rd party fonts, and this PoC/solution should be maintained as smoothly as possible in the future (e.g, the fewer the various third-party dependencies, the better).

**Comment by Vlada3333** [comment]
Interesting, thanks!
